src/ray/util/spdlog_newliner_sink.h

```diff
@@ -0,0 +1,83 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <spdlog/sinks/base_sink.h>
+
+#include <iostream>
+#include <string_view>
+
+#include "absl/strings/str_split.h"
+#include "ray/util/compat.h"
+#include "ray/util/util.h"
+
+namespace ray {
+
+// A sink which streams only after a newliner or explicit flush.
+template <typename Mutex>
+class spdlog_newliner_sink final : public spdlog::sinks::base_sink<Mutex> {
+ public:
+  explicit spdlog_newliner_sink(spdlog::sink_ptr internal_sink)
+      : internal_sink_(std::move(internal_sink)) {}
+
+ protected:
+  void sink_it_(const spdlog::details::log_msg &msg) override {
+    if (msg.payload.size() == 0) {
+      return;
+    }
+
+    const std::string_view new_content{msg.payload.data(), msg.payload.size()};
+    auto pos = new_content.find('\n');
+    if (pos == std::string_view::npos) {
+      buffer_ += new_content;
+      return;
+    }
+    std::vector<std::string_view> segments = absl::StrSplit(new_content, '\n');
+    for (int idx = 0; idx < static_cast<int>(segments.size()) - 1; ++idx) {
+      std::string cur_message = std::move(buffer_);
+      buffer_.clear();
```

Comment:
is it safe to call clear on a moved string and use it later? Should we do `buffer_ = ""` instead?

Reply:
> is it safe to call clear on a moved string and use it later?

What do you mean? I didn't access it's value. `clear` is a well-defined function.